pylint/checkers/non_ascii_names.py

```diff
@@ -0,0 +1,264 @@
+                " for a background why this could be bad."
+            ),
+            {"old_names": [("C0144", "non-ascii-name")]},
+        ),
+        # First %s will always be "file"
+        "W2402": (
+            '%s name "%s" contains a non-ASCII character. PEP 3131 only allows non-ascii identifiers!',
+            "non-ascii-file-name",
+            (
+                # Some = PyCharm at the time of writing didn't display the non_ascii_name_loł
+                # files. Probably only a bug shows the problem quite good.
+                # That's also why this is a warning and not only a convention!
+                "Some editors don't support non-ASCII file names properly. "
+                "Even so Python supports UTF-8 files since Python 3.5 this isn't recommended for"
+                "interoperability. Further reading: \n"
+                "- https://www.python.org/dev/peps/pep-0489/#export-hook-name \n"
+                "- https://www.python.org/dev/peps/pep-0672/#confusable-characters-in-identifiers\n"
+                "- https://bugs.python.org/issue20485\n"
+            ),
+        ),
+        # First %s will always be "module"
+        "C2403": (
+            '%s name "%s" contains a non-ASCII character, use an ASCII-only alias for import.',
+            "non-ascii-module-import",
+            (
+                "Used when the name contains at least one non-ASCII unicode character. "
+                "See https://www.python.org/dev/peps/pep-0672/#confusable-characters-in-identifiers"
+                " for a background why this could be bad."
+            ),
+        ),
+    }
+
+    name = "NonASCII-Checker"
+
+    def __init__(self, linter: PyLinter) -> None:
+        super().__init__(linter)
+        self._non_ascii_rgx_compiled = re.compile("[^A-Za-z0-9_]")
+
+    def _raise_name_warning(
+        self,
+        node: nodes.NodeNG,
+        node_type: str,
+        name: str,
+    ) -> None:
+        type_label = HUMAN_READABLE_TYPES.get(node_type, node_type)
+        args = (type_label.capitalize(), name)
+
+        msg = "non-ascii-name"
+
+        # Some node types have customized messages
+        if node_type == "file":
+            msg = "non-ascii-file-name"
+        elif node_type == "module":
+            msg = "non-ascii-module-import"
+
+        self.add_message(msg, node=node, args=args, confidence=interfaces.HIGH)
+
+    # pylint: disable-next=arguments-renamed
+    def _check_name(
+        self,
+        node_type: str,
+        name: str,
+        node: Union[nodes.NodeNG, _AsciiOnlyCheckedNode],
+        check_string: Optional[str] = None,
+    ) -> None:
+        """Check whether a name is using non-ASCII characters.
+
+        Also set the dynamic attribute ``_is_ascii_only`` that is used to
+        determine if a node has been already checked, so we don't have to handle
+        too many edge cases.
+        """
+
+        current_state = getattr(node, "_is_ascii_only", True)
```

Comment:
The lru_cache idea would work only if this is always the same node_type and as I can rember this is not always the case. One are assigments (=variable) and the other are the arguments of the function call.

The only idea would be to keep a list of nodes already checked. 
This would be bad memory wise i.e. once we use daemon etc.

So currently I really don't know a better way :( (besides modifying astroid to add an extra arguments `checked_by`)

Reply:
Fixed 🎉